### PR TITLE
Add tutorial for dog breed classification via ResNet-50 transfer learning

### PR DESCRIPTION
**Pull request by yushinna**
・犬の画像から犬種を判定するモデル
・ResNet-50の転移学習
・ノートブックのVersion 1はABEJA PlatformのDatasetからデータを取得して学習
・Version2はホームデリレクトリにデータをダウンロードして学習

### CODE REVIEW

**Comment by TakumiIwata** [comment]
Jupyterのソースレベルでコメント書くと場所探すのがツライので、
ノートブックのセル単位でコメントしますね

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+    "%cd '/mnt/notebooks/{notebook_id}/Platform_handson/dog_breed_classification'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# import data from ABEJA Platform Dataset\n",
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: セル２つ目
datasetIDとかのオーガナイズのIDが現在ハードコードなので
ここはマスクするか、変数として取ってくるように書き換える

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dataset_list = dataset.dataset_items.list(prefetch=False)\n",
+    "\n",
+    "dataset_list_img_label = []\n",
+    "\n",
+    "for item in tqdm(dataset_list):\n",
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: セル３
（任意）dataset_itemsの型がabeja.datasets.dataset_item.DatasetItemIterator
なので、プログレスバーではなく、作業状況のみを知らせるものになっている。
時間あれば、総数を取れる型を使ったloop変えて、プログレスバーのほうが嬉しいかも

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+    "\n",
+    "# print number of images in dataset\n",
+    "print('There are %d total dog images.' % len(dataset_list_img_label))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 2: Import library"
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: セル４まえ
(任意)dataset_list_img_label[0][0]とかdataset_list_img_label[0][1]を表示させるのは
dataset_list_img_labelできた時点でもいいかも（処理した時点で説明したほうがスムース）
その際、SDK使うとこんな感じでデータ入ると説明入れたほうがよさげ
後にdataset.props['categories'][0]['labels']も見せてもいいかも

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def image_to_tensor(img):\n",
+    "    '''\n",
+    "    As per Pytorch documentations: All pre-trained models expect input images normalized in the same way, \n",
+    "    i.e. mini-batches of 3-channel RGB images\n",
+    "    of shape (3 x H x W), where H and W are expected to be at least 224. \n",
+    "    The images have to be loaded in to a range of [0, 1] and \n",
+    "    then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225]. \n",
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: ノーマライズ向けの値結構出るので、この辺の注釈も入れてもいいかも
https://pytorch.org/docs/master/torchvision/models.html

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 5]
```diff
@@ -0,0 +1,747 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# train the model\n",
+    "n_epochs = 5\n",
+    "train(n_epochs, trainloader, validloader, model, optimizer, criterion, use_cuda, './model/dog_model.pt')"
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: モデルフォルダないよって怒られちゃうの作ること必須

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 6]
```diff
@@ -0,0 +1,747 @@
...
+    " 'ESKIMO_DOG',\n",
+    " 'FLAT-COATED_RETRIEVER',\n",
+    " 'FRENCH_BULLDOG',\n",
+    " 'GERMAN_SHEPHERD',\n",
+    " 'GERMAN_SHORT-HAIRED_POINTER',\n",
+    " 'GIANT_SCHNAUZER',\n",
+    " 'GOLDEN_RETRIEVER',\n",
+    " 'GORDON_SETTER',\n",
+    " 'GREAT_DANE',\n",
+    " 'GREAT_PYRENEES',\n",
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: カテゴリ120行は結構ボリューミーなので、改行なしで定義しちゃって
一覧などは説明資料につけるようにしたほうがよいかも

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb (hunk 5 above)
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: あと、もしハンズオン用であるなら枚数少し減らしてもいいかも。20000枚＆5エポックでも結構時間かかるので。
あと、これも進捗として、tqdm入れてほしいです。それなりの回線でも接続が切れることはあるので
処理が動いているのかないのかも知るために必要かと思います。

File: dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb [hunk 7]
```diff
@@ -0,0 +1,747 @@
...
+    "def train(n_epochs, trainloader, validloader, model, optimizer, criterion, use_cuda, save_path):\n",
+    "    \"\"\"returns trained model\"\"\"\n",
+    "    # initialize tracker for minimum validation loss\n",
+    "    valid_loss_min = 3.877533 #np.Inf\n",
+    "    \n",
+    "    if os.path.exists(save_path):\n",
+    "        model.load_state_dict(torch.load(save_path))\n",
+    "    \n",
+    "    for epoch in range(1, n_epochs+1):\n",
+    "        # initialize variables to monitor training, validation loss and validation accuracy\n",
```
Comment (TakumiIwata) [inline dog_breed_classification/notebook/Dog_breed_classification_v1.ipynb]: あと動作のソースの内容について一通り見ましたが、Pytorchお得意な人間ではないので、可読性を上げるとか
内容について効率化を求めるなら、他の人にも軽くみてもらってください！

**Comment by TakumiIwata** [comment]
@yushinna  パーッとみました。どちらかというとハンズオンするにあたっての
内容で指摘入れます。IDのハードコード削除とmodelフォルダ存在確認ができる内容が追加できれば
ご自身でマージいただいて問題ありません。その他、空き時間に必要であれば対応お願いします
